docs: Update web API doc (#11824)
unoplatform/uno: doc/articles/howto-consume-webservices.md

> [!TIP]
> For a step-by-step guide to installing the prerequisites, see [Getting started on Visual Studio](https://platform.uno/docs/articles/get-started-vs-2022.html)

## Task 1 - Create a simple Uno application

In this task, you will create a simple Single Page App with the Uno Platform. This app - **TheCatApiClient** - will be used to demonstrate an approach to consuming REST web services using the **HttpClient** class.

1. Open Visual Studio and click on **Create a new project**.

    ![Visual Studio new project dialog](Assets/tutorial01/newproject1.PNG)
1. In the **Search for Templates** search box, enter **Uno**

    ![Visual Studio new project dialog searching for Uno](Assets/tutorial01/newproject2.PNG)

1. In the filtered list of templates, select **Uno Platform App** and then click **Next**.

1. In the **Configure your new project** window, set the **Project name** to **TheCatApiClient**, choose where you would like to save your project, and click the **Create** button.

    ![Configure your new project dialog](Assets/how-to-webservice/newproject3.PNG)

> [!IMPORTANT]

> > The C# and XAML snippets in this tutorial require that the solution is named **TheCatApiClient**. Using a different name will result in build errors when you copy code from this tutorial into the solution.

Reviewer: The de-indentation change is wrong. Please indent back

* Building a web service base class that supports REST service operations and then build a number of services that derived from it.
* Using the services in a view-model.
* Build a XAML UI that utilizes the view-model.
* Test the application in Windows and WASM.

> Throughout the how-to there will be notes on recommended practices and tips that highlight resources for additional learning.

Reviewer 2: ```suggestion
Throughout the how-to, there will be notes on recommended practices and tips that highlight resources for additional learning.
```

> [!TIP]
> The complete source code that goes along with this tutorial is available in the [unoplatform/Uno.Samples](https://github.com/unoplatform/Uno.Samples) GitHub repository - [TheCatApiClient](https://github.com/unoplatform/Uno.Samples/tree/master/UI/TheCatApiClient)

> [!TIP]
> For a step-by-step guide to installing the prerequisites, see [Getting started on Visual Studio](https://platform.uno/docs/articles/get-started-vs-2022.html)

## Task 1 - Create a simple Uno application

> In this task you will create a simple Single Page App with the Uno Platform. This app - **TheCatApiClient** - will be used to demonstrate an approach to consuming REST web services using the **HttpClient** class.

Reviewer 2: ```suggestion
In this task, you will create a simple Single Page App with the Uno Platform. This app - **TheCatApiClient** - will be used to demonstrate an approach to consuming REST web services using the **HttpClient** class.
```

1. Open Visual Studio and click on **Create a new project**.

    ![Visual Studio new project dialog](Assets/tutorial01/newproject1.PNG)
1. In the **Search for Templates** search box, enter **Uno**

    ![Visual Studio new project dialog searching for Uno](Assets/tutorial01/newproject2.PNG)

1. In the filtered list of templates, select **Uno Platform App** and then click **Next**.

> 1. In the **Configure your new project** window, set the **Project name** to **TheCatApiClient**, choose where you would like to save your project and click the **Create** button.

Reviewer 2: ```suggestion
1. In the **Configure your new project** window, set the **Project name** to **TheCatApiClient**, choose where you would like to save your project, and click the **Create** button.
```

    ![Configure your new project dialog](Assets/how-to-webservice/newproject3.PNG)

> > [!IMPORTANT]

Reviewer 2: Before these other steps, we are missing the Wizard section on what options to choose for this tutorial

Author: @agneszitte Thanks! It's done

## Prerequisites

* [Visual Studio 2022 17.4 or later](https://www.visualstudio.com/downloads/)
    * **.NET Multi-platform App UI development** workload installed (for iOS, Android, Mac Catalyst development).
    * **.NET desktop development** workload installed (for Gtk, Wpf, and Linux Framebuffer development)
    * **ASP.NET** and **web development** workload installed (for WebAssembly development)

>     * [Uno Platform Extension](https://marketplace.visualstudio.com/items?itemName=unoplatform.uno-platform-addin-2022) installed

Reviewer 3: We probably should be removing that section altogether and keep the note below. It will frequently be obsolete.

    * **Uno.Wasm.Bootstrap**
    * **Uno.Wasm.Bootstrap.DevServer**

> [!IMPORTANT]

> > **Do not** update the **Microsoft.Extensions.Logging.Console**. Recent versions of the package use APIs that aren't supported by WebAssembly, and aren't compatible with Uno Platform.

Reviewer: The de-indentation change is wrong. Please indent back

1. To update the packages, click **Update**.

> 1. At the top-left of page, to browse the available packages, click **Browse**.

Reviewer 2: ```suggestion
1. At the top-left of the page, to browse the available packages, click **Browse**.
```

1. In the **Search** field, enter **System.Text.Json** and select **System.Text.Json**.

> [!TIP]

> > The **System.Text.Json** package provides functionality for serializing to and deserializing from JavaScript Object Notation (JSON). You can learn more about how to serialize and deserialize here: [How to serialize and deserialize (marshal and unmarshal) JSON in .NET](https://docs.microsoft.com/dotnet/standard/serialization/system-text-json-how-to)

Reviewer: The de-indentation change is wrong. Please indent back

    ]
    ```

> [!TIP]

> > You can learn more about the breed search API here - [GET/breeds/search](https://docs.thecatapi.com/api-reference/breeds/breeds-search).

Reviewer: The de-indentation change is wrong. Please indent back

> Now you have signed up for the API, you are ready to start implementing the API client by create a data model.

Reviewer 2: ```suggestion
Now you have signed up for the API, you are ready to start implementing the API client by creating a data model.
```

    }
    ```

> [!TIP]

> > This class is using the **JsonPropertyNameAttribute** from the **System.Text.Json** package to map the lowercase JSON property names to the [PascalCase](https://techterms.com/definition/pascalcase) convention used in C#. To learn more about **System.Text.Json**, you can review the documentation here - [How to serialize and deserialize (marshal and unmarshal) JSON in .NET](https://docs.microsoft.com/dotnet/standard/serialization/system-text-json-how-to).

Reviewer: The de-indentation change is wrong. Please indent back

You have now created a simple data model that can be used to deserialize the JSON returned by the breed search API. Next, you will create a service that interacts with the web service API.

## Task 4 - Create simple GET service for breed search

> In this task, you will create a number of classes that demonstrate how to use the **HttpClient** class to interact with a web service. Along the way, you will see some  patterns that may assist in the reuse of the code in additional projects.

Reviewer 2: ```suggestion
In this task, you will create a number of classes that demonstrate how to use the **HttpClient** class to interact with a web service. Along the way, you will see some patterns that may assist in the reuse of the code in additional projects.
```

    Notice that the **_client** variable is declared as **static**. This means there will only be one instance of the **HttpClient** shared by every instance of **WebApiBase** as **HttpClient** is intended to be instantiated once and re-used throughout the life of an application.

> [!TIP]

> > You can review the Microsoft remarks on **HttpClient** here - [HttpClient Remarks](https://docs.microsoft.com/dotnet/api/system.net.http.httpclient#remarks).

Reviewer: The de-indentation change is wrong. Please indent back

    }
    ```

    Notice that this method expects the **HttpMethod** to represent a GET, DELETE, UPDATE, or POST operation, the string URL, and an optional dictionary of headers. In this initial implementation of the **WebApiBase** class, you will consume this from the **GetAsync** method you are about to implement.

> [!NOTE]

> > You will use the header dictionary to supply the authentication key you created earlier.

Reviewer: The de-indentation change is wrong. Please indent back

    The remainder of the method uses the single **HttpClient** instance to send the **request** and `await` a response. If the response is successful, the response content is read and returned as a string. If, for some reason, the response does not indicate success, a `null` is returned.

> [!IMPORTANT]

> > As implementing adequate error handling would add many lines of code, more robust error handling is omitted for clarity. If you are interested in learning more about the considerations for handling errors with **HttpClient**, review the following blog post - [HttpClient - Error handling, a test driven approach](https://josef.codes/httpclient-error-handling-a-test-driven-approach/).

Reviewer 2: ```suggestion
> As implementing adequate error handling would add many lines of code, more robust error handling is omitted for clarity. If you are interested in learning more about the considerations for handling errors with **HttpClient**, review the following blog post - [HttpClient - Error handling, a test-driven approach](https://josef.codes/httpclient-error-handling-a-test-driven-approach/).
```

> > As implementing adequate error handling would add many lines of code, more robust error handling is omitted for clarity. If you are interested in learning more about the considerations for handling errors with **HttpClient**, review the following blog post - [HttpClient - Error handling, a test-driven approach](https://josef.codes/httpclient-error-handling-a-test-driven-approach/).

Reviewer: The de-indentation change is wrong. Please indent back

    First, you will note that the method is asynchronous. You will then notice the use of string interpolation and the **WebUtility.HtmlEncode** method to construct the API url. Also note how the request headers are supplied to the call - the **accept** header advises the API that the client expects JSON, and the **x-api-key** specifies your API key.

> [!IMPORTANT]

> > Replace the `{YOUR-API-KEY}` value with the API key you were sent when you registered with TheCatApi.

Reviewer: The de-indentation change is wrong. Please indent back

Once the result is returned, it is null-checked and, if not-null, the JSON is deserialized into an enumerable collection of **Breed** instances using the data model you created earlier. If the result is null, then an empty list is returned.

At this point, you have implemented a simple breed search service. Now, it is time to create the view and view-model that will allow a user to enter a search term and display the results.

## Task 5 - Create a base view model

In this sample application, you will be adopting the Model-View-ViewModel (MVVM) pattern, which helps to cleanly separate business and presentation logic. In this task, you will be creating the base view model that will be used in the following task to create the view model for the main page.

> [!TIP]

> > If you want to learn more about MVVM, review the document here - [The Model-View-ViewModel Pattern](https://docs.microsoft.com/xamarin/xamarin-forms/enterprise-application-patterns/mvvm).

Reviewer: ```suggestion
> If you want to learn more about MVVM, review the [The Model-View-ViewModel Pattern documentation](https://learn.microsoft.com/xamarin/xamarin-forms/enterprise-application-patterns/mvvm).
```

    ```csharp
    // Insert the event and property below here
    public event PropertyChangedEventHandler PropertyChanged;
    protected DispatcherQueue Dispatcher => DispatcherQueue.GetForCurrentThread();
    ```

    The **PropertyChanged** event declaration satisfies the **INotifyPropertyChanged** interface contract, but isn't particularly helpful alone - you will add methods that make it easier to use shortly.

    The **Dispatcher** property is used later to ensure property updates are performed on the UI thread - an essential detail for updating controls in multi-tasking applications.

> [!TIP]

> > You can learn more about the UI Thread and multi-tasking here - [Keep the UI thread responsive](https://docs.microsoft.com/windows/uwp/debug-test-perf/keep-the-ui-thread-responsive)

Reviewer: The de-indentation change is wrong. Please indent back

    The method body then compares the incoming value with the current property value, and returns false if they are equal. This prevents a situation where a property is updated, which updates a control via binding, which then sets the same property with the same value again. Without this guard, the code would continue to loop.

    The **value** is then assigned to the **backingVariable** (updating the variable in the caller). The property changed event is raised (you will add that next) and a true value is returned.

> [!TIP]

> > You can learn more about **CallerMemberNameAttribute** here - [Reserved attributes: Determine caller information](https://docs.microsoft.com/dotnet/csharp/language-reference/attributes/caller-information).

Reviewer: The de-indentation change is wrong. Please indent back

>     The **value** is then assigned to the **backingVariable** (updating the variable in the caller). The property changed event is raise (you will add that next) and a true value is returned.

Reviewer 2: ```suggestion
    The **value** is then assigned to the **backingVariable** (updating the variable in the caller). The property changed event is raised (you will add that next) and a true value is returned.
```

    }
    ```

    This method uses a helper method (added next) to ensure the event is raised on the UI thread. As the code does not await this execution, the compiler would usually display a warning with the assumption that this is a mistake. As it is intentional, the code includes `#pragma` directives that temporarily disable that warning.

> [!TIP]

> > If you want to learn more about lambda expressions, review the document here - [Lambda expressions](https://docs.microsoft.com/dotnet/csharp/language-reference/operators/lambda-expressions)

Reviewer: The de-indentation change is wrong. Please indent back

>     This method uses a helper method (added next) to ensure the event is raised on the UI thread. As the code does not await this execution, the compiler would usually display a warning in the assumption that this is a mistake. As it is intentional, the code includes `#pragma` directives that temporarily disable that warning.

Reviewer 2: ```suggestion
    This method uses a helper method (added next) to ensure the event is raised on the UI thread. As the code does not await this execution, the compiler would usually display a warning with the assumption that this is a mistake. As it is intentional, the code includes `#pragma` directives that temporarily disable that warning.
```

        }
        else
        {
            var completion = new TaskCompletionSource();

>             Dispatcher.TryEnqueue(()=>

Reviewer 3: ```suggestion
            Dispatcher.TryEnqueue(() =>
```

            {
                callback();
                completion.SetResult();
            });
            await completion.Task; 
        }
    }
    ```

> [!NOTE]

> > As **WASM** is currently single-threaded, the **Uno.WASM** implementation of `DispatcherQueue.HasThreadAccess` always returns `false` by default. This default has been chosen to prevent live-locking for certain frameworks that expect a multi-threaded environment. This app overrides that default WASM behavior here via conditional compilation and always invokes the callback directly on WASM.

Reviewer: The de-indentation change is wrong. Please indent back

> > As **WASM** is currently single-threaded, the **Uno.WASM** implementation of `DispatcherQueue.HasThreadAccess` always returns `false` by default. This default has been chosen to prevent live-locking for certain frameworks that expect a multi-threaded environment. This app overrides that default WASM behavior here via conditional compilation and always invokes the callback directly on WASM.

Reviewer: @jeromelaban I'm not sure if this is still true or if it is outdated statement. Can you confirm?

Reviewer 3: This is still true. There's no direct API to determine if threading is supported (something that .NET 8 will provide).

> * In XAML applications, control bindings should only be updated on the UI thread, therefore this code uses the value of **hasThreadAccess** to determine if the callback can be directly invoked. If not, the **DispatcherQueue.TryEnqueue** method is used to execute the callback on the UI Thread. This method is used when setting properties and whenever bound collections are updated, to ensure the collection views (such as GridView, ListView, etc.) are updated.

Reviewer: The de-indentation change is wrong. Please indent back

    #pragma warning restore CS4014 // Because this call is not awaited, execution of the current method continues before the call is completed
            }

        // Insert DispatchAsync below here
        protected async Task DispatchAsync(DispatcherQueueHandler callback)
        {
            var hasThreadAccess =
        #if __WASM__
            true;
        #else
            Dispatcher.HasThreadAccess;
        #endif

            if (hasThreadAccess)
            {
                callback.Invoke();
            }
            else
            {
                var completion = new TaskCompletionSource();

>                 Dispatcher.TryEnqueue(()=>

Reviewer 3: ```suggestion
                Dispatcher.TryEnqueue(() =>
```

> In this task you will build the view-model that implements a simple breed search feature using the base class and web service client you just implemented.

Reviewer 2: ```suggestion
In this task, you will build the view-model that implements a simple breed search feature using the base class and web service client you just implemented.
```

    Notice that this grid is displayed from row 1 and spans 4 rows. As this grid is defined last, it will appear on top of the other controls, block click/touch interactions with the underlying controls as well showing a semi-opaque message that the app is **Downloading data**.

    ![Windows app Search UI preview](Assets/how-to-webservice/uwp-ui-busy.png)

> [!NOTE]

> > You would typically use a **ProgressRing** here, but in a effort to keep the UI as simple as possible, this approach was taken.

Reviewer 2: ```suggestion
> You would typically use a **ProgressRing** here, but in an effort to keep the UI as simple as possible, this approach was taken.
```

> > You would typically use a **ProgressRing** here, but in an effort to keep the UI as simple as possible, this approach was taken.

Reviewer: The de-indentation change is wrong. Please indent back

    This service implements two methods - **GetByBreed** which returns back all images for a breed and **GetById**, which returns a specific image. They each utilize the **WebApiBase.GetAsync** method you used earlier.

> [!TIP]

> > You can review the images service documentation here - [/images](https://docs.thecatapi.com/api-reference/images/)

Reviewer: The de-indentation change is wrong. Please indent back

    }
    ```

    These methods are very similar to the methods you implemented earlier - **GetAll** returns a collection of the users **Favorite** instances, whereas **Get** returns a single **Favorite** by its ID.

> [!TIP]

> > You can review the favorites service documentation here - [/favorites](https://docs.thecatapi.com/api-reference/favourites/)

Reviewer: The de-indentation change is wrong. Please indent back

>     These methods are very similar to the methods you have implemented earlier - **GetAll** returns a collection of the users **Favorite** instances, whereas **Get** returns a single **Favorite** by its ID.

Reviewer 2: ```suggestion
    These methods are very similar to the methods you implemented earlier - **GetAll** returns a collection of the users **Favorite** instances, whereas **Get** returns a single **Favorite** by its ID.
```

> In this final task you will bring together the new **FavoritesApi** with the updated **MainViewModel** to add a list of favorites to the UI.

Reviewer 2: ```suggestion
In this final task, you will bring together the new **FavoritesApi** with the updated **MainViewModel** to add a list of favorites to the UI.
```

    The **Downloading data...** UI should be shown and then the results should be populated. You should be able to scroll the results and select one (nothing will happen yet).

    ![Windows app Search UI preview](Assets/how-to-webservice/uwp-ui-preview.png)

1. In the search results, click **Abyssinian** and you should see an image of a cat added to the **Favorites** list.

    ![Windows app Search UI preview](Assets/how-to-webservice/client-with-favorite.png)

1. Repeat the above, adding images for other breeds.

1. In the **Favorites** list, click an image and you should see it be removed.

1. Ensure you have one or more images remaining in your **Favorites** list and exit the Windows application.

1. To test the app in Wasm, in the **Solution Explorer**, right-click the **TheCatApiClient.Wasm** project and select **Set as Startup project**.

1. To run the application, press **F5**.

    Visual Studio will open a browser and display the application. You should see the same favorites you had listed in the Windows app.

1. Search for additional breeds and add and remove favorites - you should see the same functionality as the Windows app.

> 1. Try the app on the other platforms. Below is an image of app running on the iPhone and Android simulators.

Reviewer 2: ```suggestion
1. Try the app on other platforms. Below is an image of the app running on iPhone and Android simulators.
```

>     ![app running on iPhone and Android simulators](Assets/how-to-webservice/favorites-on-phone.png)

Reviewer 2: ```suggestion
    ![Application running on iPhone and Android simulators](Assets/how-to-webservice/favorites-on-phone.png)
```
